java-checks/src/main/java/org/sonar/java/checks/security/SMTPSSLServerIdentityCheck.java

```diff
@@ -0,0 +1,139 @@
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ConstantUtils;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4499")
+public class SMTPSSLServerIdentityCheck extends AbstractMethodDetection {
+
+  private static final String APACHE = "org.apache.commons.mail.Email";
+  private static final String BOOLEAN = "boolean";
+  private static final String HASHTABLE = "java.util.Hashtable";
+
+  private static final MethodMatcher SET_SSL_ON_CONNECT = MethodMatcher.create()
+    .typeDefinition(TypeCriteria.is(APACHE))
+    .name("setSSLOnConnect")
+    .addParameter(BOOLEAN);
+
+  private static final MethodMatcher HASHTABLE_PUT = MethodMatcher.create()
+    .typeDefinition(TypeCriteria.is(HASHTABLE))
+    .name("put")
+    .withAnyParameters();
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(SET_SSL_ON_CONNECT, HASHTABLE_PUT);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    MethodTree method = findEnclosingMethod(mit);
+    if (method != null) {
+      Arguments args = mit.arguments();
+      if (SET_SSL_ON_CONNECT.matches(mit) && LiteralUtils.isTrue(args.get(0))) {
+        MethodBodyApacheVisitor apacheVisitor = new MethodBodyApacheVisitor();
+        method.accept(apacheVisitor);
+        if (!apacheVisitor.isSecured) {
+          reportIssue(mit, "Enable server identity validation on this SMTP SSL connection.");
+        }
+      } else if (HASHTABLE_PUT.matches(mit) && parametersSocketFactoryMatch(args.get(0), args.get(1))) {
+        MethodBodyHashtableVisitor hashVisitor = new MethodBodyHashtableVisitor();
+        method.accept(hashVisitor);
+        if (!hashVisitor.isSecured) {
+          reportIssue(mit, "Enable server identity validation, set \"mail.smtp.ssl.checkserveridentity\" to true");
+        }
+      }
+    }
+    super.onMethodInvocationFound(mit);
+  }
+
+  private static boolean argIsNotBooleanOrIsTrue(ExpressionTree arg) {
+    return LiteralUtils.isTrue(arg) || !arg.is(Tree.Kind.BOOLEAN_LITERAL);
+  }
+
+  @CheckForNull
+  private static MethodTree findEnclosingMethod(Tree tree) {
+    while (!tree.is(Tree.Kind.CLASS, Tree.Kind.METHOD)) {
+      tree = tree.parent();
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      return null;
+    }
+    return (MethodTree) tree;
+  }
+
+  private static boolean parametersSocketFactoryMatch(ExpressionTree arg1, ExpressionTree arg2) {
```

Comment:
Having a separate method for that does not seem really useful and the method signature does not convey its intent.